File: backendOLD/runpod-api/queue-manager.ts
import dotenv from 'dotenv';
import path from 'path';
import fs from 'fs';

// Load environment variables
dotenv.config({ path: path.resolve(process.cwd(), '.env') });
dotenv.config({ path: path.resolve(process.cwd(), '.env.local') });

import { recruitPod, pausePod, resumePod, releasePod, getPodStatus } from './index';
import { getRunpodGraphQLClient } from './client';

// Load configuration
const configPath = path.resolve(process.cwd(), 'backend', 'runpod-api', 'config.json');
const config = JSON.parse(fs.readFileSync(configPath, 'utf8'));

export interface WorkflowRequest {
  id: string;
  workflowName: string;
  requestData: any;
  timestamp: number;
  status: 'pending' | 'processing' | 'completed' | 'failed';
  podId?: string;
}

export interface ActivePod {
  id: string;
  workflowName: string;
  createdAt: number;
  lastUsedAt: number;
  pauseTimeoutAt: number;
  terminateTimeoutAt: number;
  pausedAt?: number;
  status: 'running' | 'paused' | 'terminated';
  requestQueue: WorkflowRequest[];
}

export interface QueueManager {
  activePods: Map<string, ActivePod>;
  pendingRequests: Map<string, WorkflowRequest[]>;
  isRunning: boolean;
  intervalId?: NodeJS.Timeout;
}

class WorkflowQueueManager {
  private queueManager: QueueManager;
  private client: any;

  constructor() {
    this.queueManager = {
      activePods: new Map(),
      pendingRequests: new Map(),
      isRunning: false
    };
    this._client = null;
  }

  private getClient() {
    if (!this._client) {
      this._client = getRunpodGraphQLClient();
    }
    return this._client;
  }

  private _client: any = null;

  async start() {
    if (this.queueManager.isRunning) {
      console.log('⚠️ Queue manager is already running');
      return;
    }

    console.log('🚀 Starting Workflow Queue Manager...');
    this.queueManager.isRunning = true;

    // Start the main processing loop
    this.queueManager.intervalId = setInterval(() => {
      this.processQueue();
    }, config.queueSettings.checkInterval);

    console.log('✅ Queue manager started successfully');
  }

  async stop() {
    if (!this.queueManager.isRunning) {
      console.log('⚠️ Queue manager is not running');
      return;
    }

    console.log('🛑 Stopping Workflow Queue Manager...');
    this.queueManager.isRunning = false;

    if (this.queueManager.intervalId) {
      clearInterval(this.queueManager.intervalId);
      this.queueManager.intervalId = undefined;
    }

    // Pause all active pods
    for (const [podId, pod] of this.queueManager.activePods) {
      if (pod.status === 'running') {
        console.log(`⏸️ Pausing pod ${podId} before shutdown`);
        await pausePod(podId);
      }
    }

    console.log('✅ Queue manager stopped successfully');
  }

  async addWorkflowRequest(workflowName: string, requestData: any): Promise<string> {
    const requestId = `req_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
    
    const request: WorkflowRequest = {
      id: requestId,
      workflowName: workflowName.toLowerCase(),
      requestData,
      timestamp: Date.now(),
      status: 'pending'
    };

    // Add to pending requests
    if (!this.queueManager.pendingRequests.has(request.workflowName)) {
      this.queueManager.pendingRequests.set(request.workflowName, []);
    }
    this.queueManager.pendingRequests.get(request.workflowName)!.push(request);

    console.log(`📝 Added workflow request: ${request.workflowName} (${requestId})`);
    
    // Try to process immediately
    await this.processQueue();

    return requestId;
  }

  private async processQueue() {
    if (!this.queueManager.isRunning) return;

    // Process each workflow type
    for (const [workflowName, requests] of this.queueManager.pendingRequests) {
      if (requests.length === 0) continue;

      // Find available pod for this workflow
      let availablePod = this.findAvailablePod(workflowName);
      
      if (!availablePod) {
        // Check if we're already creating a pod for this workflow
        const creatingPod = Array.from(this.queueManager.activePods.values())
          .find(pod => pod.workflowName === workflowName);
        
        if (creatingPod) {
          console.log(`⏳ Pod ${creatingPod.id} is already being created for workflow ${workflowName}`);
          continue; // Skip this workflow for now
        }
        
        // Try to create a new pod
        availablePod = await this.createPodForWorkflow(workflowName);
      }

      if (availablePod) {
        // Process requests for this pod
        await this.processPodRequests(availablePod);
      }
    }

    // Check for pods that need to be paused due to timeout
    await this.checkPodTimeouts();
  }

  private findAvailablePod(workflowName: string): ActivePod | null {
    for (const [podId, pod] of this.queueManager.activePods) {
      if (pod.workflowName === workflowName && 
          (pod.status === 'running' || pod.status === 'paused') && 
          pod.requestQueue.length < 3) { // Max 3 concurrent requests per pod
        return pod;
      }
    }
    return null;
  }

  private async createPodForWorkflow(workflowName: string): Promise<ActivePod | null> {
    try {
      console.log(`🏗️ Creating new pod for workflow: ${workflowName}`);
      
      // Check if we're already creating a pod for this workflow
      const existingPod = Array.from(this.queueManager.activePods.values())
        .find(pod => pod.workflowName === workflowName && pod.status === 'running');
      
      if (existingPod) {
        console.log(`⚠️ Pod already exists for workflow ${workflowName}: ${existingPod.id}`);
        return existingPod;
      }
      
      // Check network volume availability first
      const hasNetworkVolumes = await this.checkNetworkVolumeAvailability();
      if (!hasNetworkVolumes) {
        console.log('⚠️ No network volumes available, creating pod without network volume');
      }
      
      const podConfig = {
        name: `${workflowName}-pod-${Date.now()}`,
        imageName: config.podSettings.defaultImage,
        cloudType: 'SECURE' as const, // Use SECURE for network volume access
        gpuCount: config.podSettings.defaultGpuCount,
        minMemoryInGb: config.podSettings.defaultMemoryInGb,
        countryCode: 'SE', // Use Sweden to match network volume data center
        supportPublicIp: config.podSettings.supportPublicIp,
        containerDiskInGb: config.podSettings.defaultDiskInGb,
        minVcpuCount: config.podSettings.defaultVcpuCount,
        ports: config.podSettings.defaultPorts,
        dockerArgs: '',
        maxRetries: 2, // Reduced retries to prevent multiple pods
        retryDelay: 3000, // Reduced delay
        workflowName: workflowName // Pass workflow name for network volume selection
      };

      console.log(`🎯 Pod config for ${workflowName}:`, {
        name: podConfig.name,
        imageName: podConfig.imageName,
        cloudType: podConfig.cloudType,
        countryCode: podConfig.countryCode,
        workflowName: podConfig.workflowName,
        hasNetworkVolumes: hasNetworkVolumes
      });

      const result = await recruitPod(podConfig);
      
      if (result.success && result.pod) {
        const timeouts = this.getWorkflowTimeouts(workflowName);
        const now = Date.now();
        const activePod: ActivePod = {
          id: result.pod.id,
          workflowName,
          createdAt: now,
          lastUsedAt: now,
          pauseTimeoutAt: now + (timeouts.pause * 1000),
          terminateTimeoutAt: now + (timeouts.terminate * 1000),
          status: 'running',
          requestQueue: []
        };

        this.queueManager.activePods.set(result.pod.id, activePod);
        console.log(`✅ Created pod ${result.pod.id} for workflow ${workflowName}`);
        
        return activePod;
      } else {
        console.log(`❌ Failed to create pod for workflow ${workflowName}: ${result.error}`);
        return null;
      }
    } catch (error) {
      console.log(`❌ Error creating pod for workflow ${workflowName}:`, error);
      return null;
    }
  }

  private async processPodRequests(pod: ActivePod) {
    const pendingRequests = this.queueManager.pendingRequests.get(pod.workflowName) || [];
    
    if (pendingRequests.length === 0) return;

    // Move requests to pod queue
    const requestsToProcess = pendingRequests.splice(0, 3 - pod.requestQueue.length);
    pod.requestQueue.push(...requestsToProcess);

    console.log(`🔄 Queuing ${requestsToProcess.length} requests for pod ${pod.id}`);

    // Mark requests as processing and assign to pod
    for (const request of requestsToProcess) {
      request.status = 'processing';
      request.podId = pod.id;
      console.log(`📋 Request ${request.id} assigned to pod ${pod.id} for ${request.workflowName} workflow`);
    }

    // Update pod last used time and reset timeouts
    const now = Date.now();
    pod.lastUsedAt = now;
    const timeouts = this.getWorkflowTimeouts(pod.workflowName);
    pod.pauseTimeoutAt = now + (timeouts.pause * 1000);
    pod.terminateTimeoutAt = now + (timeouts.terminate * 1000);
    
    // If pod was paused, resume it
    if (pod.status === 'paused') {
      try {
        const resumeResult = await resumePod(pod.id);
        if (resumeResult.success) {
          pod.status = 'running';
          pod.pausedAt = undefined;
          console.log(`🔄 Successfully resumed pod ${pod.id} due to new requests`);
        } else {
          console.log(`❌ Failed to resume pod ${pod.id}: ${resumeResult.error}`);
          return; // Don't process requests if we can't resume the pod
        }
      } catch (error) {
        console.log(`❌ Error resuming pod ${pod.id}:`, error);
        return; // Don't process requests if we can't resume the pod
      }
    }
  }



  private async waitForPodWithIp(podId: string, maxAttempts: number = 12): Promise<{ success: boolean; pod?: any; error?: string }> {
    console.log(`⏳ Waiting for pod ${podId} to have IP address...`);
    
    for (let attempt = 1; attempt <= maxAttempts; attempt++) {
      try {
        const podStatus = await getPodStatus(podId);
        
        if (podStatus.success && podStatus.data) {
          const pod = podStatus.data as any;
          const status = pod.desiredStatus || pod.status;
          console.log(`📊 Pod status (attempt ${attempt}/${maxAttempts}): ${status}`);
          
          if (status === 'RUNNING') {
            // Try to get the pod's public IP
            const podWithIp = await this.getPodPublicIp(podId);
            if (podWithIp.success && podWithIp.pod?.ip) {
              console.log(`✅ Pod is ready with IP: ${podWithIp.pod.ip}`);
              return podWithIp;
            } else {
              console.log(`⏳ Pod is running but IP not yet available, waiting...`);
            }
          } else if (status === 'FAILED' || status === 'TERMINATED' || status === 'EXITED') {
            return {
              success: false,
              error: `Pod failed with status: ${status}`
            };
          }
        }
        
        // Wait 10 seconds before next attempt
        await new Promise(resolve => setTimeout(resolve, 10000));
        
      } catch (error) {
        console.log(`❌ Error checking pod status (attempt ${attempt}):`, error);
        await new Promise(resolve => setTimeout(resolve, 5000));
      }
    }
    
    return {
      success: false,
      error: `Pod did not become ready with IP within ${maxAttempts * 10} seconds`
    };
  }

  private async getPodPublicIp(podId: string): Promise<{ success: boolean; pod?: any; error?: string }> {
    try {
      // Use the REST API to get pod details
      const client = this.getClient();
      const podResult = await client.getPodById(podId);
      
      if (podResult.success && podResult.data) {
        const pod = podResult.data as any;
        
        // Check if the pod has a public IP (it might be empty initially)
        const publicIp = pod.publicIp;
        if (publicIp && publicIp !== '') {
          return {
            success: true,
            pod: { ...pod, ip: publicIp }
          };
        } else {
          return {
            success: false,
            error: 'Pod does not have a public IP address yet'
          };
        }
      } else {
        return {
          success: false,
          error: podResult.error || 'Failed to get pod details'
        };
      }
    } catch (error) {
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Unknown error getting pod IP'
      };
    }
  }


  private async checkPodTimeouts() {
    const now = Date.now();
    
    for (const [podId, pod] of this.queueManager.activePods) {
      if (pod.status === 'running' && 
          pod.requestQueue.length === 0 && 
          now > pod.pauseTimeoutAt) {
        
        console.log(`⏰ Pod ${podId} reached pause timeout, pausing...`);
        pod.status = 'paused';
        pod.pausedAt = now;
        
        try {
          await pausePod(podId);
          console.log(`✅ Pod ${podId} paused due to timeout`);
        } catch (error) {
          console.log(`❌ Failed to pause pod ${podId}:`, error);
        }
      } else if (pod.status === 'paused' && 
                 pod.pausedAt && 
                 now > pod.terminateTimeoutAt) {
        
        console.log(`⏰ Pod ${podId} reached terminate timeout, terminating...`);
        pod.status = 'terminated';
        
        try {
          await releasePod(podId);
          console.log(`✅ Pod ${podId} terminated due to timeout`);
          this.queueManager.activePods.delete(podId);
        } catch (error) {
          console.log(`❌ Failed to terminate pod ${podId}:`, error);
        }
      }
    }
  }

  private getWorkflowTimeouts(workflowName: string): { pause: number; terminate: number } {
    return config.workflow[workflowName]?.timeouts || config.workflow.default.timeouts;
  }

  private getWorkflowNetworkVolume(workflowName: string): string {
    return config.workflow[workflowName]?.["network-volume"] || config.workflow.default["network-volume"];
  }

  private async checkNetworkVolumeAvailability(): Promise<boolean> {
    try {
      console.log('🔍 Checking network volume availability...');
      const client = this.getClient();
      const volumesResult = await client.getNetworkVolumes();
      
      if (volumesResult.success && volumesResult.data && volumesResult.data.length > 0) {
        console.log(`✅ Found ${volumesResult.data.length} available network volumes`);
        volumesResult.data.forEach((volume: any, index: number) => {
          console.log(`   ${index + 1}. ${volume.name} (${volume.id}) - ${volume.size}GB - ${volume.dataCenterId}`);
        });
        return true;
      } else {
        console.log('❌ No network volumes available');
        return false;
      }
    } catch (error) {
      console.log('❌ Error checking network volumes:', error);
      return false;
    }
  }

  getQueueStatus() {
    return {
      activePods: Array.from(this.queueManager.activePods.values()),
      pendingRequests: Object.fromEntries(
        Array.from(this.queueManager.pendingRequests.entries()).map(([name, requests]) => [
          name, 
          requests.filter(req => req.status === 'pending')
        ])
      ),
      isRunning: this.queueManager.isRunning
    };
  }

  getPodForWorkflow(workflowName: string): ActivePod | null {
    return this.findAvailablePod(workflowName);
  }

  getPodById(podId: string): ActivePod | null {
    return this.queueManager.activePods.get(podId) || null;
  }

  async getPodWithIp(podId: string): Promise<{ success: boolean; pod?: any; error?: string }> {
    return await this.waitForPodWithIp(podId);
  }

  markRequestCompleted(requestId: string, result?: any) {
    for (const [podId, pod] of this.queueManager.activePods) {
      const request = pod.requestQueue.find(req => req.id === requestId);
      if (request) {
        request.status = 'completed';
        if (result) {
          (request as any).result = result;
        }
        console.log(`✅ Request ${requestId} marked as completed on pod ${podId}`);
        return true;
      }
    }
    return false;
  }

  markRequestFailed(requestId: string, error?: string) {
    for (const [podId, pod] of this.queueManager.activePods) {
      const request = pod.requestQueue.find(req => req.id === requestId);
      if (request) {
        request.status = 'failed';
        if (error) {
          (request as any).error = error;
        }
        console.log(`❌ Request ${requestId} marked as failed on pod ${podId}`);
        return true;
      }
    }
    return false;
  }

  async cleanup() {
    console.log('🧹 Cleaning up queue manager...');
    
    // Terminate all active pods
    for (const [podId, pod] of this.queueManager.activePods) {
      try {
        await releasePod(podId);
        console.log(`✅ Terminated pod ${podId}`);
      } catch (error) {
        console.log(`❌ Failed to terminate pod ${podId}:`, error);
      }
    }

    this.queueManager.activePods.clear();
    this.queueManager.pendingRequests.clear();
  }
}

// Singleton instance
let queueManagerInstance: WorkflowQueueManager | null = null;

export function getQueueManager(): WorkflowQueueManager {
  if (!queueManagerInstance) {
    queueManagerInstance = new WorkflowQueueManager();
  }
  return queueManagerInstance;
}

export { WorkflowQueueManager };

File: api/runpod/queue_manager.py
# queue_manager.py
# FastAPI-ready port of your TypeScript WorkflowQueueManager
# ----------------------------------------------------------
from __future__ import annotations

import asyncio
import json
import os
from dataclasses import dataclass, field, asdict
from pathlib import Path
from typing import Any, Dict, List, Literal, Optional, Tuple

from dotenv import load_dotenv
from fastapi import APIRouter, FastAPI, HTTPException
from pydantic import BaseModel, Field

# --- env + config ------------------------------------------------------------

CWD = Path(os.getcwd())
load_dotenv(CWD / ".env")
load_dotenv(CWD / ".env.local", override=True)

CONFIG_PATH = CWD / "config.json"
if not CONFIG_PATH.exists():
    raise RuntimeError(f"Missing config file at {CONFIG_PATH}")

with CONFIG_PATH.open("r", encoding="utf-8") as f:
    CONFIG: Dict[str, Any] = json.load(f)

# --- external deps: adapt these imports to your project ----------------------
# Expect these async helpers to exist in your codebase.
# They should mirror your Node functions' behavior.
#
# recruitPod(pod_config) -> dict {success: bool, pod?: {id: str}, error?: str}
# pausePod(pod_id) -> dict {success: bool, error?: str}
# resumePod(pod_id) -> dict {success: bool, error?: str}
# releasePod(pod_id) -> dict {success: bool, error?: str}
# getPodStatus(pod_id) -> dict {success: bool, data?: {...}, error?: str}
# getRunpodGraphQLClient() -> object with:
#   - getPodById(pod_id) -> {success: bool, data?: {...}, error?: str}
#   - getNetworkVolumes() -> {success: bool, data?: [ {...} ], error?: str}

from client import get_runpod_graphql_client, get_runpod_rest_client

# --- types -------------------------------------------------------------------

StatusReq = Literal["pending", "processing", "completed", "failed"]
StatusPod = Literal["running", "paused", "terminated"]

@dataclass
class WorkflowRequest:
    id: str
    workflowName: str
    requestData: Any
    timestamp: int
    status: StatusReq
    podId: Optional[str] = None
    result: Optional[Any] = None
    error: Optional[str] = None


@dataclass
class ActivePod:
    id: str
    workflowName: str
    createdAt: int
    lastUsedAt: int
    pauseTimeoutAt: int
    terminateTimeoutAt: int
    status: StatusPod
    requestQueue: List[WorkflowRequest] = field(default_factory=list)
    pausedAt: Optional[int] = None


class QueueStatus(BaseModel):
    activePods: List[Dict[str, Any]]
    pendingRequests: Dict[str, List[Dict[str, Any]]]
    isRunning: bool


class AddWorkflowBody(BaseModel):
    workflowName: str = Field(..., description="Name of the workflow")
    requestData: Any = Field(..., description="Arbitrary request payload")


class MarkBody(BaseModel):
    result: Optional[Any] = None
    error: Optional[str] = None


# --- manager -----------------------------------------------------------------

class WorkflowQueueManager:
    def __init__(self) -> None:
        self.activePods: Dict[str, ActivePod] = {}
        self.pendingRequests: Dict[str, List[WorkflowRequest]] = {}
        self.isRunning: bool = False
        self._task: Optional[asyncio.Task] = None
        self._lock = asyncio.Lock()
        self._client = None

        qs = CONFIG.get("queueSettings", {})
        self.check_interval_ms: int = int(qs.get("checkInterval", 2000))

        ps = CONFIG.get("podSettings", {})
        self.default_image = ps.get("defaultImage")
        self.default_gpu = ps.get("defaultGpuCount", 1)
        self.default_mem_gb = ps.get("defaultMemoryInGb", 8)
        self.default_vcpu = ps.get("defaultVcpuCount", 4)
        self.default_disk_gb = ps.get("defaultDiskInGb", 20)
        self.support_public_ip = bool(ps.get("supportPublicIp", True))
        self.default_ports = ps.get("defaultPorts", [])

    # --- client ---------------------------------------------------------------

    def _get_client(self):
        if self._client is None:
            self._client = get_runpod_graphql_client()
        return self._client

    # --- lifecycle ------------------------------------------------------------

    async def start(self) -> None:
        async with self._lock:
            if self.isRunning:
                return
            self.isRunning = True
            self._task = asyncio.create_task(self._loop(), name="workflow-queue-loop")

    async def stop(self) -> None:
        async with self._lock:
            if not self.isRunning:
                return
            self.isRunning = False

        if self._task:
            self._task.cancel()
            try:
                await self._task
            except asyncio.CancelledError:
                pass
            self._task = None

        # pause all running pods
        for pod_id, pod in list(self.activePods.items()):
            if pod.status == "running":
                try:
                    client = get_runpod_rest_client()
                    await client.pause_pod(pod_id)
                except Exception:
                    pass

    async def cleanup(self) -> None:
        # terminate all pods
        for pod_id in list(self.activePods.keys()):
            try:
                client = get_runpod_rest_client()
                await client.terminate_pod(pod_id)
            except Exception:
                pass
        self.activePods.clear()
        self.pendingRequests.clear()

    # --- public API -----------------------------------------------------------

    async def add_workflow_request(self, workflow_name: str, request_data: Any) -> str:
        import time, random, string

        workflow_name = workflow_name.lower()
        rid = f"req_{int(time.time()*1000)}_" + "".join(
            random.choices(string.ascii_lowercase + string.digits, k=9)
        )
        req = WorkflowRequest(
            id=rid,
            workflowName=workflow_name,
            requestData=request_data,
            timestamp=int(time.time() * 1000),
            status="pending",
        )
        async with self._lock:
            self.pendingRequests.setdefault(workflow_name, []).append(req)
        # nudge the loop without waiting for the next tick
        asyncio.create_task(self._process_queue_once())
        return rid

    def get_queue_status(self) -> QueueStatus:
        pending: Dict[str, List[Dict[str, Any]]] = {}
        for name, reqs in self.pendingRequests.items():
            pending[name] = [asdict(r) for r in reqs if r.status == "pending"]

        return QueueStatus(
            activePods=[asdict(p) for p in self.activePods.values()],
            pendingRequests=pending,
            isRunning=self.isRunning,
        )

    def get_pod_for_workflow(self, workflow_name: str) -> Optional[ActivePod]:
        return self._find_available_pod(workflow_name.lower())

    def get_pod_by_id(self, pod_id: str) -> Optional[ActivePod]:
        return self.activePods.get(pod_id)

    async def get_pod_with_ip(self, pod_id: str, max_attempts: int = 12) -> Dict[str, Any]:
        return await self._wait_for_pod_with_ip(pod_id, max_attempts)

    def mark_request_completed(self, request_id: str, result: Any = None) -> bool:
        for pod in self.activePods.values():
            for req in pod.requestQueue:
                if req.id == request_id:
                    req.status = "completed"
                    req.result = result
                    return True
        return False

    def mark_request_failed(self, request_id: str, error: Optional[str] = None) -> bool:
        for pod in self.activePods.values():
            for req in pod.requestQueue:
                if req.id == request_id:
                    req.status = "failed"
                    req.error = error
                    return True
        return False

    # --- internal loop --------------------------------------------------------

    async def _loop(self) -> None:
        try:
            while self.isRunning:
                await self._process_queue_once()
                await asyncio.sleep(self.check_interval_ms / 1000.0)
        except asyncio.CancelledError:
            pass

    async def _process_queue_once(self) -> None:
        if not self.isRunning:
            return

        async with self._lock:
            workflow_items = list(self.pendingRequests.items())

        # process each workflow type
        for workflow_name, requests in workflow_items:
            if not requests:
                continue

            pod = self._find_available_pod(workflow_name)
            if not pod:
                # check if one exists/being created
                creating = next(
                    (p for p in self.activePods.values() if p.workflowName == workflow_name),
                    None,
                )
                if creating:
                    continue
                pod = await self._create_pod_for_workflow(workflow_name)

            if pod:
                await self._process_pod_requests(pod)

        await self._check_pod_timeouts()

    # --- helpers --------------------------------------------------------------

    def _find_available_pod(self, workflow_name: str) -> Optional[ActivePod]:
        for pod in self.activePods.values():
            if (
                pod.workflowName == workflow_name
                and pod.status in ("running", "paused")
                and len(pod.requestQueue) < 3  # max 3 concurrent requests
            ):
                return pod
        return None

    def _get_workflow_timeouts(self, workflow_name: str) -> Tuple[int, int]:
        wf = CONFIG.get("workflow", {})
        wfc = wf.get(workflow_name, wf.get("default", {}))
        to = wfc.get("timeouts", {})
        return int(to.get("pause", 60)), int(to.get("terminate", 300))

    def _get_workflow_network_volume(self, workflow_name: str) -> Optional[str]:
        wf = CONFIG.get("workflow", {})
        wfc = wf.get(workflow_name, wf.get("default", {}))
        return wfc.get("network-volume")

    async def _check_network_volume_availability(self) -> bool:
        try:
            client = get_runpod_rest_client()
            res = await client.getNetworkVolumes()
            return bool(res.success and res.data)
        except Exception:
            return False

    async def _create_pod_for_workflow(self, workflow_name: str) -> Optional[ActivePod]:
        import time

        # network volume check (non-fatal)
        has_net_vol = await self._check_network_volume_availability()

        pod_config = {
            "name": f"{workflow_name}-pod-{int(time.time()*1000)}",
            "imageName": self.default_image,
            "cloudType": "SECURE",  # to match TS
            "gpuCount": self.default_gpu,
            "minMemoryInGb": self.default_mem_gb,
            "countryCode": "SE",
            "supportPublicIp": self.support_public_ip,
            "containerDiskInGb": self.default_disk_gb,
            "minVcpuCount": self.default_vcpu,
            "ports": self.default_ports,
            "dockerArgs": "",
            "maxRetries": 2,
            "retryDelay": 3000,
            "workflowName": workflow_name,
            "hasNetworkVolumes": has_net_vol,
        }

        client = get_runpod_rest_client()
        result = await client.createPod(pod_config)
        if result.get("success") and result.get("pod"):
            now = int(time.time() * 1000)
            pause_s, term_s = self._get_workflow_timeouts(workflow_name)
            active = ActivePod(
                id=result["pod"]["id"],
                workflowName=workflow_name,
                createdAt=now,
                lastUsedAt=now,
                pauseTimeoutAt=now + pause_s * 1000,
                terminateTimeoutAt=now + term_s * 1000,
                status="running",
            )
            self.activePods[active.id] = active
            return active
        return None

    async def _process_pod_requests(self, pod: ActivePod) -> None:
        import time

        # move up to 3 - current queued
        capacity = 3 - len(pod.requestQueue)
        if capacity <= 0:
            return

        async with self._lock:
            pending = self.pendingRequests.get(pod.workflowName, [])
            to_process = pending[:capacity]
            self.pendingRequests[pod.workflowName] = pending[capacity:]

        if not to_process:
            return

        # mark + assign
        for req in to_process:
            req.status = "processing"
            req.podId = pod.id
        pod.requestQueue.extend(to_process)

        # update timers
        now = int(time.time() * 1000)
        pod.lastUsedAt = now
        pause_s, term_s = self._get_workflow_timeouts(pod.workflowName)
        pod.pauseTimeoutAt = now + pause_s * 1000
        pod.terminateTimeoutAt = now + term_s * 1000

        # resume if paused
        if pod.status == "paused":
            try:
                client = get_runpod_rest_client()
                r = await client.startPod(pod.id)
                if r.get("success"):
                    pod.status = "running"
                    pod.pausedAt = None
                else:
                    # cannot process if can't resume
                    return
            except Exception:
                return

    async def _wait_for_pod_with_ip(self, pod_id: str, max_attempts: int = 12) -> Dict[str, Any]:
        for attempt in range(1, max_attempts + 1):
            try:
                client = get_runpod_rest_client()
                status = await client.getPodById(pod_id)
                if status.get("success") and status.get("data"):
                    pod_data = status["data"]
                    pod_state = pod_data.get("desiredStatus") or pod_data.get("status")
                    if pod_state == "RUNNING":
                        got = await self._get_pod_public_ip(pod_id)
                        if got.get("success") and got.get("pod", {}).get("ip"):
                            return got
                # FAILED / TERMINATED / EXITED
                if status.get("data", {}).get("status") in {"FAILED", "TERMINATED", "EXITED"}:
                    return {"success": False, "error": f"Pod failed with status: {status['data']['status']}"}
            except Exception:
                pass
            await asyncio.sleep(10)
        return {"success": False, "error": f"Pod did not become ready with IP within {max_attempts * 10} seconds"}

    async def _get_pod_public_ip(self, pod_id: str) -> Dict[str, Any]:
        try:
            client = get_runpod_rest_client()
            res = await client.getPodById(pod_id)
            if res.get("success") and res.get("data"):
                pod = res["data"]
                ip = pod.get("publicIp")
                if ip:
                    pod_copy = dict(pod)
                    pod_copy["ip"] = ip
                    return {"success": True, "pod": pod_copy}
                return {"success": False, "error": "Pod does not have a public IP address yet"}
            return {"success": False, "error": res.get("error") or "Failed to get pod details"}
        except Exception as e:
            return {"success": False, "error": str(e) or "Unknown error getting pod IP"}

    async def _check_pod_timeouts(self) -> None:
        import time

        now = int(time.time() * 1000)
        for pod_id, pod in list(self.activePods.items()):
            if pod.status == "running" and len(pod.requestQueue) == 0 and now > pod.pauseTimeoutAt:
                # pause
                pod.status = "paused"
                pod.pausedAt = now
                try:
                    await pausePod(pod_id)
                except Exception:
                    pass
            elif pod.status == "paused" and pod.pausedAt and now > pod.terminateTimeoutAt:
                # terminate
                pod.status = "terminated"
                try:
                    await releasePod(pod_id)
                except Exception:
                    pass
                self.activePods.pop(pod_id, None)


# --- singleton accessor -------------------------------------------------------

_manager_instance: Optional[WorkflowQueueManager] = None

def get_queue_manager() -> WorkflowQueueManager:
    global _manager_instance
    if _manager_instance is None:
        _manager_instance = WorkflowQueueManager()
    return _manager_instance


# --- FastAPI router -----------------------------------------------------------

router = APIRouter(prefix="/queue", tags=["queue"])

@router.post("/start")
async def start_queue():
    mgr = get_queue_manager()
    await mgr.start()
    return {"ok": True}

@router.post("/stop")
async def stop_queue():
    mgr = get_queue_manager()
    await mgr.stop()
    return {"ok": True}

@router.post("/request")
async def add_request(body: AddWorkflowBody):
    mgr = get_queue_manager()
    rid = await mgr.add_workflow_request(body.workflowName, body.requestData)
    return {"ok": True, "requestId": rid}

@router.get("/status", response_model=QueueStatus)
async def status():
    return get_queue_manager().get_queue_status()

@router.get("/pods/{pod_id}")
async def get_pod(pod_id: str):
    pod = get_queue_manager().get_pod_by_id(pod_id)
    if not pod:
        raise HTTPException(404, "Pod not found")
    return asdict(pod)

@router.get("/pods/{pod_id}/ip")
async def wait_for_ip(pod_id: str, max_attempts: int = 12):
    res = await get_queue_manager().get_pod_with_ip(pod_id, max_attempts=max_attempts)
    if not res.get("success"):
        raise HTTPException(400, res.get("error", "Failed"))
    return res

@router.post("/requests/{request_id}/complete")
async def mark_complete(request_id: str, body: MarkBody):
    ok = get_queue_manager().mark_request_completed(request_id, result=body.result)
    if not ok:
        raise HTTPException(404, "Request not found")
    return {"ok": True}

@router.post("/requests/{request_id}/fail")
async def mark_fail(request_id: str, body: MarkBody):
    ok = get_queue_manager().mark_request_failed(request_id, error=body.error)
    if not ok:
        raise HTTPException(404, "Request not found")
    return {"ok": True}

@router.post("/cleanup")
async def cleanup():
    await get_queue_manager().cleanup()
    return {"ok": True}

# Optional: mount directly if you want a self-contained app file.
def create_app() -> FastAPI:
    app = FastAPI(title="Workflow Queue Manager")
    app.include_router(router)
    return app

app = create_app()
